source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/UtiliOS.mm

```diff
@@ -656,3 +656,25 @@ void TextRunToRichTextElementProperties(const std::shared_ptr<TextRun> &textRun,
     textProp.SetItalic(textRun->GetItalic());
     textProp.SetStrikethrough(textRun->GetStrikethrough());
 }
+
+ACOBaseActionElement *deserializeUnknowActionToCustomAction(const std::shared_ptr<UnknownAction> unknownAction)
+{
+    ACRRegistration *reg = [ACRRegistration getInstance];
+    ACOBaseActionElement *customAction = nil;
+    if (reg) {
+        NSString *type = [NSString stringWithCString:unknownAction->GetElementTypeString().c_str() encoding:NSUTF8StringEncoding];
+        NSObject<ACOIBaseActionElementParser> *parser = [reg getCustomActionElementParser:type];
+        if (!parser) {
+            @throw [ACOFallbackException fallbackException];
+        }
+        Json::Value blob = unknownAction->GetAdditionalProperties();
+        Json::FastWriter fastWriter;
+        NSString *jsonString = [[NSString alloc] initWithCString:fastWriter.write(blob).c_str() encoding:NSUTF8StringEncoding];
+        if (jsonString.length > 0) {
+            NSData *jsonPayload = [jsonString dataUsingEncoding:NSUTF8StringEncoding];
+            ACOParseContext *context = [reg getParseContext];
+            customAction = [parser deserialize:jsonPayload parseContext:context];
```

Comment:
>deserialize:jsonPayload [](start = 35, length = 23)

It's unfortunate that you have to serialize and then immediately deserialize this JSON payload, because, I guess, the C++ core and the Obj-C layers are using different JSON libraries. I don't know enough about the AdaptiveCards architecture to suggest whether there's any way to avoid this, though.

Reply:
When shared model parses json using JSON library we use, it converts json to Json::Value, our parser walks the Json:Value returned, and create & initialize adaptive objects. So we have to find a way to avoid parsing json string to Json::Value during parsing custom type at the shared model, and it's out of scope for this fix entails.


---
In reply to: [338183418](https://github.com/microsoft/AdaptiveCards/pull/3526#discussion_r338183418) [](ancestors = 338183418)